_bleio/packet_buffer.py

```diff
@@ -0,0 +1,119 @@
+# The MIT License (MIT)
+#
+# Copyright (c) 2020 Dan Halbert for Adafruit Industries LLC
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""
+`_bleio.packet_buffer`
+=======================================================================
+
+_bleio implementation for Adafruit_Blinka_bleio
+
+* Author(s): Dan Halbert for Adafruit Industries
+"""
+import queue
+from typing import Union
+
+from _bleio import Characteristic
+
+Buf = Union[bytes, bytearray, memoryview]
+
+
+class PacketBuffer:
+    """Accumulates a Characteristic's incoming packets in a FIFO buffer and facilitates packet aware
+    outgoing writes. A packet's size is either the characteristic length or the maximum transmission
+    unit (MTU) minus overhead, whichever is smaller. The MTU can change so check
+    `incoming_packet_length` and `outgoing_packet_length` before creating a buffer to store data.
+
+    When we're the server, we ignore all connections besides the first to subscribe to
+    notifications."""
+
+    def __init__(self, characteristic: Characteristic, *, buffer_size: int):
+        """Monitor the given Characteristic. Each time a new value is written to the Characteristic
+        add the newly-written bytes to a FIFO buffer.
+
+        :param Characteristic characteristic: The Characteristic to monitor.
+          It may be a local Characteristic provided by a Peripheral Service,
+          or a remote Characteristic in a remote Service that a Central has connected to.
+        :param int buffer_size: Size of ring buffer (in packets of the Characteristic's maximum
+          length) that stores incoming packets coming from the peer."""
+        self._queue = queue.Queue(buffer_size)
```

Comment:
Don't you need two queues? One incoming and one outgoing?

Reply:
@tannewt and I ended up talking about this. I can't do an outgoing queue because I can't tell that there's a pending packet still waiting to be sent. So all the packets are just sent one at at time: no concatenation is possible.